app/Config.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+module Config
+  ( Config (..)
+  , CopyThing (..) )
+where
+
+import Data.Aeson
+import Data.Maybe (fromMaybe)
+import Data.Yaml
+import Path
+import System.Hapistrano.Commands
+
+-- | Hapistrano configuration typically loaded from @hap.yaml@ file.
+
+data Config = Config
+  { configDeployPath :: !(Path Abs Dir)
+    -- ^ Top-level deploy directory on target machine
+  , configHost :: !(Maybe String)
+    -- ^ Host to deploy to. If missing, localhost will be assumed.
+  , configPort :: !(Maybe Word)
+    -- ^ SSH port number to use, may be omitted
+  , configRepo :: !String
+    -- ^ Location of repository that contains the source code to deploy
+  , configRevision :: !String
+    -- ^ Revision to use
+  , configRestartCommand :: !(Maybe GenericCommand)
+    -- ^ The command to execute when switching to a different release
+    -- (usually after a deploy or rollback).
+  , configBuildScript :: !(Maybe [GenericCommand])
+    -- ^ Build script to execute to build the project
+  , configCopyFiles :: ![CopyThing]
+    -- ^ Collection of files to copy over to target machine before building
+  , configCopyDirs :: ![CopyThing]
+    -- ^ Collection of directories to copy over to target machine before building
+  } deriving (Eq, Ord, Show)
+
+-- | Information about source and destination locations of a file\/directory
+-- to copy.
+
+data CopyThing = CopyThing FilePath FilePath
+  deriving (Eq, Ord, Show)
+
+instance FromJSON Config where
+  parseJSON = withObject "Hapistrano configuration" $ \o -> do
+    configDeployPath <- o .: "deploy_path"
+    configHost       <- o .:? "host"
+    configPort       <- o .:? "port"
```

Comment:
@mrkkrp what do you think about setting the default port value at this level:

```Haskell
o .:? "port" .!= 22
```

That way we don't have to deal with the `Maybe` elsewhere.